hbase-client/src/main/java/org/apache/hadoop/hbase/client/RawAsyncTableImpl.java

```diff
@@ -796,52 +797,102 @@ private boolean locateFinished(RegionInfo region, byte[] endKey, boolean endKeyI
 
   private <S, R> void coprocessorServiceUntilComplete(Function<RpcChannel, S> stubMaker,
     ServiceCaller<S, R> callable, PartialResultCoprocessorCallback<S, R> callback,
-    AtomicBoolean locateFinished, AtomicInteger unfinishedRequest, RegionInfo region, Span span) {
+    MultiRegionCoprocessorServiceProgress<R> progress, RegionInfo region, Span span) {
     addListener(coprocessorService(stubMaker, callable, region, region.getStartKey()), (r, e) -> {
       try (Scope ignored = span.makeCurrent()) {
-        if (e != null) {
-          callback.onRegionError(region, e);
-        } else {
-          callback.onRegionComplete(region, r);
-        }
+        if (e instanceof RegionNameChangedException) {
+          RegionInfo newRegion = ((RegionNameChangedException)e).getNewRegionInfo();
+          if (progress.markNewRegionAndCheckNeedsToBeHandled(newRegion)) {
+            if (progress.hasResponseStarted(region)) {
+              // already started sending responses for the original region, so we can't switch to
+              // the new regions because that would violate the contract of the callbacks
+              progress.onResponse(region, null, new DoNotRetryIOException(
+                "Region " + region.getEncodedName() + " no longer exists, likely due to a split or"
+                  + " merge, and the coprocessor service was already in progress and can't be recovered"
+              ));
```

Comment:
In this case, the `CoprocessorCallback::onRegionComplete` has already been called at least once for this region, and I think it'd be better to throw an error in this edge case instead of trying to make the `CoprocessorCallback` handle this scenario. In this example scenario, if we had already sent these callbacks:
```
callback.onRegionComplete(r1, { sum => 10, next => 0x01 });
callback.onRegionComplete(r1, { sum => 7, next => 0x02 });
[r1 is split into r2 and r3]
```
How would we signal to the caller that the results for region r1 are actually no longer relevant, and that they should now consider those results replaced by r2 and r3? Again, I think keeping the interface as is and continuing to throw an error in this edge case would be better, but open to other thoughts